api/v1/controllers/collectors.js

```diff
@@ -362,75 +362,78 @@ function heartbeat(req, res, next) {
  * @returns {Object} - Response of the start endpoint
  */
 function startCollector(req, res, next) {
   const resultObj = { reqStartTime: req.timestamp };
-  const requestBody = req.swagger.params.queryBody.value;
-  requestBody.status = 'Running';
-  requestBody.createdBy = req.user.id;
-  let collectorToReturn;
-  return helper.model.findOne({ where: { name: requestBody.name } })
-    .then((collector) => {
-      if (collector) {
-        return u.isWritable(req, collector);
+  const body = req.swagger.params.queryBody.value;
+  body.status = 'Running';
+  body.createdBy = req.user.id;
+  let collToReturn;
+  let name;
+  return helper.model.findOne({ where: { name: body.name } })
+  /* Already exists? Verify that this user has write permission. */
+  .then((coll) => {
+    if (coll) return u.isWritable(req, coll);
+    return coll;
+  })
+  /* Already exists and is running or paused? Error! */
+  .then((coll) => {
+    if (coll) {
+      if (coll.status === 'Running' || coll.status === 'Paused') {
+        throw new apiErrors.ForbiddenError({
+          explanation: 'Cannot start--only a stopped collector can start',
+        });
       }
 
-      return collector;
-    })
-    .then((collector) => {
-      if (collector) {
-        if (!collector.registered) {
-          throw new apiErrors.ForbiddenError({ explanation:
-            'Cannot start--this collector is not registered.',
-          });
-        }
-
-        if (collector.status === 'Running' || collector.status === 'Paused') {
-          throw new apiErrors.ForbiddenError({ explanation:
-            'Cannot start--only a stopped collector can start',
-          });
-        }
+      if (!coll.registered) {
+        throw new apiErrors.ForbiddenError({
+          explanation: 'You must reregister this collector before you can ' +
+            'start it',
+        });
       }
+    }
 
-      return collector;
-    })
-    .then((collector) => collector ? collector.update(requestBody) :
-      helper.model.create(requestBody))
-    .then((collector) => {
-      collectorToReturn = collector;
-      return collector.getCurrentGenerators();
-    })
-    .then((generators) => Promise.all(generators.map((g) =>
-      g.updateForHeartbeat())))
-    .then((generators) => {
-      resultObj.dbTime = new Date() - resultObj.reqStartTime;
-      collectorToReturn.dataValues.generatorsAdded = generators.map((g) => {
-        g.token = jwtUtil.createToken(g.user.name, g.user.name);
-        delete g.GeneratorCollectors;
-        delete g.collectors;
-        return g;
-      });
+    return coll;
+  })
+  /* Update or create */
+  .then((coll) => coll ? coll.update(body) : helper.model.create(body))
+  .then((coll) => {
+    name = coll.name;
+    collToReturn = coll;
+    return coll.getCurrentGenerators();
+  })
+  /* Filter: only want generators marked as "current" for this collector. */
+  .then((gens) => gens.filter((g) => g.currentCollector = name))
```

Comment:
Shouldn't coll.getCurrentGenerators already be returning the correct list of current generators?

Reply:
Yeah, that's in the "trackGenerators" story -- this filter shouldn't be necessary if that is working.

Reply:
So this is just a stop-gap until that gets fixed?

Reply:
Er, I wasn't thinking of it that way but I think you're right.

Reply:
Ok. Can you add a TODO to remove the filter once getCurrentGenerators is working?

Reply:
@iamigo can you see my comment in this old pr?

There was a typo in this line that we didn't catch in the review: "=" instead of "==". So this wasn't actually filtering. If it had been correct, the trust collector would have been broken when this went out about a month ago.

This is why I was confused earlier. I was pretty sure I hadn't made any functional changes in the reassign generators PR. I had replaced this filter with a find by collector name, which is functionally the same. But the filter wasn't actually filtering, so my change made it work correctly, which broke the generator assignment since we hadn't fully thought through how to make the transition.

I updated the next story to change this back once all the assignment is in place.

Reply:
Good catch @jgraff2